SDK 1.0 compatibility fixes

keith-packard opened the pull request:
These patches improve compiler compatibility with the upcoming SDK 1.0 which includes GCC 14.3. They are ready for review and merging.

keith-packard committed:
modules/cmsis-dsp: Don't use Zephyr stdint.h

Zephyr replaces the toolchain version of stdint.h to define uint32_t and
int32_t as int rather than long. This breaks the ARM MVE intrinics which
require uint32_t to be defined as unsigned long.

Define 'ZEPHYR_USE_TOOLCHAIN_STDINT_H_' as that causes zephyr_stdint.h to
be skipped, ensuring that the normal stdint.h types are used instead.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
arch/x86: Support picolibc with zefi

Disable the default startup file with -nostartfiles.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
soc/intel_adsp: soc_adsp_halt_cpu always fails when NUM_CPUS <= 1

When the target has only a single CPU, this function cannot ever
succeed. Skip all of the drama and just return -EINVAL. This makes GCC 14
happy as it doesn't get confused about possible out of bounds access of the
soc_cpus_active array.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
drivers/adc/adc_stm32: Check both single-ended and differential defines

The stm32u3x header files defines LL_ADC_SINGLE_ENDED but not
LL_ADC_DIFFERENTIAL as the device doesn't support differential mode. The
driver only checked for LL_ADC_SINGLE_ENDED and assumed that when that was
defined, LL_ADC_DIFFERENTIAL would also be defined.

Check for both when figuring out which calibration type will be required.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/kernel: Avoid C library calls from interrupt handlers

On arm64, GCC feels free to use floating point registers for essentially
anything unless we build with -mgeneral-regs-only. After the FPU gets used
in an interrupt handler, interrupts are disabled as there's no place to
save the registers during a nested exception. As the C library may be built
separately without this flag, we cannot use any of its functions or risk
having nested exceptions fail.

Switch printk usage to k_str_out and stick to (mostly) Zephyr internal
functions to ensure nested interrupts work correctly.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
toolchain/gcc: Add "memory" clobber to asm traps before CODE_UNREACHABLE

GCC 14.3 will happily delete any code that appears before
__builtin_unreachable that isn't separated with an obvious branch. That
includes __asm__ statements, even those which generate traps.

The failure case that I debugged was on x86 in
z_check_stack_sentinel. There is a store to restore the sentinel to the
correct value just before the ARCH_EXCEPT, and that macro emits 'int $32'
followed by CODE_UNREACHABLE. Because the compiler didn't understand that
ARCH_EXCEPT was changing execution flow, it decided that the sentinel
restoring store 'couldn't' be reached and elided it.

I added the "memory" clobber to the asm statement in ARCH_EXCEPT before
CODE_UNREACHABLE to enforce that all pending store operations be performed
before the asm statement occurs. This ensures that they are not deleted by
the compiler.

I think this might be a GCC bug. The GCC documentation explicitly documents
that asm statements which change the flow of control should be followed by
__builtin_unreachable.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/arm_mpu_pxn: Insist that trivial functions are not inlined

GCC 14 takes a look these functions which are just 'return true' and
inlines them even though they has the noinline attribute set. This happens
because the compiler computes the possible range of values from the
function call, and as that is a single value (true), it replaces the call
with that value. So it's not strictly inlining the function?

Insist a bit harder by adding a compiler_barrier to the functions which
seems to solve the issue.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/kernel: Loosen bounds on timeslice schedule for slow clocked systems

RX consistently fails this test by a tiny amount. Loosen the tolerance to
let it work.

Signed-off-by: Keith Packard <keithp@keithp.com>

peter-mitsis commented on tests/kernel/sched/schedule_api/src/test_sched_timeslice_reset.c:
```diff
@@ -28,7 +28,7 @@ BUILD_ASSERT(NUM_THREAD <= MAX_NUM_THREAD);
 #else
 /* ... 1ms is faster than a tick, loosen tolerance to 1 tick */
-#define TASK_SWITCH_TOLERANCE (1000 / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#define TASK_SWITCH_TOLERANCE (1100 / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
```
Nit: The preceding comment should be updated.

keith-packard replied:
Thanks. Updated.

peter-mitsis reviewed (dismissed):
Looks good -- just one minor nitpick.

keith-packard commented:
argh. my blanket addition of a compiler barrier before CODE_UNREACHABLE broke tests. I've fallen back to adding "memory" clobbers to a bunch of asm statements which change code flow.

keith-packard committed:
tests/mem_protect: Adjust backing store size

This test is failing; perhaps we've shrunk the kernel enough to change what
this value needs to be?

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
Had to adjust the amount of backing store for the demand paging tests.
